File: rust/skialin-core/src/path_measure.rs
use crate::{sys, Matrix, Path, PathBuilder, Point};

pub struct PosTan {
    pub position: Point,
    pub tangent: Point,
}

pub struct PathMeasure(*mut sys::SkPathMeasure);

impl PathMeasure {
    pub fn new(path: &Path, force_closed: bool, res_scale: f32) -> Self {
        PathMeasure(unsafe { sys::skialin_bridge_PathMeasure_new(path.0, force_closed, res_scale) })
    }

    /// A path measure with no path attached; call [`Self::set_path`] before using it.
    pub fn empty() -> Self {
        PathMeasure(unsafe { sys::skialin_bridge_PathMeasure_new(std::ptr::null(), false, 1.0) })
    }

    /// `path` of `None` clears the current path.
    pub fn set_path(&mut self, path: Option<&Path>, force_closed: bool) {
        let ptr = path.map_or(std::ptr::null(), |p| p.0 as *const sys::SkPath);
        unsafe { sys::skialin_bridge_PathMeasure_setPath(self.0, ptr, force_closed) };
    }

    /// The length of the current contour, or 0 if there's no path.
    pub fn length(&mut self) -> f32 {
        unsafe { sys::skialin_bridge_PathMeasure_getLength(self.0) }
    }

    /// `distance` is pinned to `[0, length]`. `None` if there's no path or it's zero-length.
    pub fn pos_tan(&mut self, distance: f32) -> Option<PosTan> {
        let mut position = sys::SkPoint::default();
        let mut tangent = sys::SkPoint::default();
        let ok = unsafe { sys::skialin_bridge_PathMeasure_getPosTan(self.0, distance, &mut position, &mut tangent) };
        ok.then_some(PosTan { position: position.into(), tangent: tangent.into() })
    }

    /// `distance` is pinned to `[0, length]`. `None` if there's no path or it's zero-length.
    pub fn matrix(&mut self, distance: f32) -> Option<Matrix> {
        let mut out = Matrix::identity();
        let ok = unsafe { sys::skialin_bridge_PathMeasure_getMatrix(self.0, distance, &mut out.0, 0x3) };
        ok.then_some(out)
    }

    /// Appends the `[start_d, stop_d]` segment of the current contour to `dst`.
    /// `false` (and `dst` untouched) if the segment is zero-length or `start_d > stop_d`.
    pub fn segment(&mut self, start_d: f32, stop_d: f32, dst: &mut PathBuilder, start_with_move_to: bool) -> bool {
        unsafe { sys::skialin_bridge_PathMeasure_getSegment(self.0, start_d, stop_d, dst.as_raw_mut(), start_with_move_to) }
    }

    pub fn is_closed(&mut self) -> bool {
        unsafe { sys::skialin_bridge_PathMeasure_isClosed(self.0) }
    }

    /// Advances to the next contour in the path. `true` if one exists.
    pub fn next_contour(&mut self) -> bool {
        unsafe { sys::skialin_bridge_PathMeasure_nextContour(self.0) }
    }
}

impl Drop for PathMeasure {
    fn drop(&mut self) {
        unsafe { sys::skialin_bridge_PathMeasure_delete(self.0) };
    }
}

File: src/main/kotlin/org/skialin/PathMeasure.kt
package org.skialin

import org.skialin.impl.Managed
import org.skialin.impl.NativeLoader

/** Measures distance along a path (length, position/tangent, matrix at a point). Mirrors Skia's `SkPathMeasure`. */
class PathMeasure(
    path: Path,
    forceClosed: Boolean = false,
    resScale: Float = 1f,
) : Managed(PathMeasureNative.nNew(path.nativePtr, forceClosed, resScale), PathMeasureNative::nRelease) {
    data class PosTan(
        val position: Point,
        val tangent: Point,
    )

    /** `path` of `null` clears the current path. */
    fun setPath(
        path: Path?,
        forceClosed: Boolean = false,
    ) = PathMeasureNative.nSetPath(nativePtr, path?.nativePtr ?: 0L, forceClosed)

    /** The length of the current contour, or 0 if there's no path. */
    fun length(): Float = PathMeasureNative.nLength(nativePtr)

    /** [distance] is pinned to `[0, length]`. `null` if there's no path or it's zero-length. */
    fun posTan(distance: Float): PosTan? {
        val out = FloatArray(4)
        if (!PathMeasureNative.nPosTan(nativePtr, distance, out)) return null
        return PosTan(Point(out[0], out[1]), Point(out[2], out[3]))
    }

    /** [distance] is pinned to `[0, length]`. `null` if there's no path or it's zero-length. */
    fun matrix(distance: Float): Matrix33? {
        val out = FloatArray(9)
        if (!PathMeasureNative.nMatrix(nativePtr, distance, out)) return null
        return Matrix33(out)
    }

    /**
     * Appends the `[startD, stopD]` segment of the current contour to [dst]. Returns false
     * ([dst] untouched) if the segment is zero-length or `startD > stopD`.
     */
    fun segment(
        startD: Float,
        stopD: Float,
        dst: PathBuilder,
        startWithMoveTo: Boolean,
    ): Boolean = PathMeasureNative.nSegment(nativePtr, startD, stopD, dst.nativePtr, startWithMoveTo)

    val isClosed: Boolean get() = PathMeasureNative.nIsClosed(nativePtr)

    /** Advances to the next contour in the path. `true` if one exists. */
    fun nextContour(): Boolean = PathMeasureNative.nNextContour(nativePtr)
}

private object PathMeasureNative {
    init {
        NativeLoader.ensureLoaded()
    }

    external fun nNew(
        pathPtr: Long,
        forceClosed: Boolean,
        resScale: Float,
    ): Long

    external fun nRelease(ptr: Long)

    external fun nSetPath(
        ptr: Long,
        pathPtr: Long,
        forceClosed: Boolean,
    )

    external fun nLength(ptr: Long): Float

    external fun nPosTan(
        ptr: Long,
        distance: Float,
        out: FloatArray,
    ): Boolean

    external fun nMatrix(
        ptr: Long,
        distance: Float,
        out: FloatArray,
    ): Boolean

    external fun nSegment(
        ptr: Long,
        startD: Float,
        stopD: Float,
        dstPtr: Long,
        startWithMoveTo: Boolean,
    ): Boolean

    external fun nIsClosed(ptr: Long): Boolean

    external fun nNextContour(ptr: Long): Boolean
}
